dataset/test/shape_test.cpp

```diff
@@ -451,3 +451,51 @@ TEST(ReshapeTest, round_trip_with_all) {
   EXPECT_EQ(flatten(reshaped, std::vector<Dim>{Dim::Row, Dim::Time}, Dim::X),
             a);
 }
+
+class TransposeTest : public ::testing::Test {
+protected:
+  TransposeTest() : a(xy) {
+    a.coords().set(Dim::X, x);
+    a.coords().set(Dim::Y, y);
+    a.masks().set("mask-x", x);
+    a.masks().set("mask-y", y);
+    a.masks().set("mask-xy", xy);
+  }
+  Variable xy = makeVariable<double>(Dims{Dim::Y, Dim::X}, Shape{3, 2},
+                                     Values{1, 2, 3, 4, 5, 6});
+  Variable x = xy.slice({Dim::Y, 0});
+  Variable y = xy.slice({Dim::X, 0});
+  DataArray a;
+};
+
+TEST_F(TransposeTest, data_array_2d) {
+  auto transposed = transpose(a);
+  EXPECT_EQ(transposed.data(), transpose(a.data()));
+  transposed.setData(a.data());
+  EXPECT_EQ(transposed, a);
+  EXPECT_EQ(transpose(a, {Dim::X, Dim::Y}), transpose(a));
+  EXPECT_EQ(transpose(a, {Dim::Y, Dim::X}), a);
+}
+
+TEST_F(TransposeTest, dataset_no_order) {
+  Dataset d;
+  d.setData("a", a);
+  d.setData("b", transpose(a));
+  d.setData("c", a.slice({Dim::X, 0}));
+  // Slightly unusual but "simple" behavior if no dim order given
+  auto transposed = transpose(d);
+  EXPECT_EQ(transposed["a"], d["b"]);
+  EXPECT_EQ(transposed["b"], d["a"]);
+  EXPECT_EQ(transposed["c"], d["c"]);
+}
+
+TEST_F(TransposeTest, dataset_2d) {
+  Dataset d;
+  d.setData("a", a);
+  d.setData("b", transpose(a));
+  auto transposed = transpose(d, {Dim::X, Dim::Y});
+  EXPECT_EQ(transposed["a"], d["b"]);
+  EXPECT_EQ(transposed["b"], d["b"]);
+  d.setData("c", a.slice({Dim::X, 0}));
+  EXPECT_THROW_DISCARD(transpose(d, {Dim::X, Dim::Y}), except::DimensionError);
+}
```

Comment:
I think we should have a test that checks the intentional behaviour that masks and 2d coords are not transposed.